compose/runtime/runtime/src/nativeMain/kotlin/androidx/compose/runtime/ActualNative.native.kt

```diff
@@ -134,13 +134,10 @@ internal actual fun <T> createSnapshotMutableState(
 ): SnapshotMutableState<T> =
     SnapshotMutableStateImpl(value, policy)
 
-// fixme: not actually thread local
 internal actual class SnapshotThreadLocal<T> actual constructor() {
-    private var value: T? = null
-
-    actual fun get(): T? = value
+    actual fun get(): T? = threadLocalStorage.get(this) as T
     actual fun set(value: T?) {
-        this.value = value
+        threadLocalStorage.set(this, value)
```

Comment:
According to the documentation of SnapshotThreadLocal avoiding weak references is expected behavior:

```
/**
 * This is similar to a [ThreadLocal] but has lower overhead because it avoids a weak reference.
 * This should only be used when the writes are delimited by a try...finally call that will clean
 * up the reference such as [androidx.compose.runtime.snapshots.Snapshot.enter] else the reference
 * could get pinned by the thread local causing a leak.
 *
 * [ThreadLocal] can be used to implement the actual for platforms that do not exhibit the same
 * overhead for thread locals as the JVM and ART.
 */

```

The fact that we do not use weak references for ThreadLocal implementation for K/N may lead to a potential problem, though your are right that unfortunately there is no weak references for K/N at all so we have to live with it until clear motivation for weak references appears for K/N .

Reply:
Apologies if I misunderstand, but it looks like there is a WeakReference class in Kotlin/Native since 1.3. There's currently no WeakHashMap that I can find (and intellij doesn't suggest it).
https://github.com/JetBrains/kotlin/blob/v1.8.0/kotlin-native/runtime/src/main/kotlin/kotlin/native/ref/Weak.kt